src/group_theory/nilpotent.lean

```diff
@@ -328,6 +336,78 @@ begin
     use [lower_central_series G, lower_central_series_is_descending_central_series, h] },
 end
 
+section classical
+
+open_locale classical
+
+variables [hG : is_nilpotent G]
+include hG
+
+/-- The nilpotency class of a nilpotent group is the small natural `n` such that
+the `n`'th term of the upper central series is `G`. -/
+noncomputable def group.nilpotency_class : ℕ :=
+nat.find (is_nilpotent.nilpotent G)
+
+/-- Equivalently, the nilpotency class is the smallest `n` for which any ascending
+central series reaches `G` in its `n`'th term -/
+lemma least_ascending_central_series_length_eq_nilpotency_class
+  (h : ∃ n : ℕ, ∃ H : ℕ → subgroup G, (is_ascending_central_series H ∧ H n = ⊤)) :
+  nat.find h = @group.nilpotency_class G _ _ :=
+begin
```

Comment:
I am not sure hwo to best express this. The `h` assumption can actually be obtained from `is_nilpotent G`, so I guess I could write something like
```suggestion
lemma least_ascending_central_series_length_eq_nilpotency_class :
  let 
    (h : ∃ n : ℕ, ∃ H : ℕ → subgroup G, (is_ascending_central_series H ∧ H n = ⊤)) := begin … end
   in nat.find h = @group.nilpotency_class G _ _ :=
begin
```
or even inline it. But it seems to obscure the lemma’s statement.

Given proof irrelevance, does it matter?

Reply:
I think I would prefer a version that is both readable and avoids the assumption. Can we do that be making some aux definition?

Reply:
I could introduce a aux definition (`has_finite_ascending_central_series`).

But once we have that, one might expect the theorems such as
```
theorem nilpotent_iff_finite_descending_central_series :
  is_nilpotent G ↔ ∃ H : ℕ → subgroup G, is_descending_central_series H ∧ ∃ n : ℕ, H n = ⊥ :=
```
to be expressed using that definition. Which would be fine with me, it’s just moving the  `∃ n` to the front (which is necessary to make it compatible with `nat.find`).

Reply:
Actually, I didn’t need new defintions, but merely had to reorder the `∃` in the statement `theorem nilpotent_iff_finite_ascending_central_series` so that its' RHS is a suitable argument to `nat.find`. What do you think of it now?

Reply:
Thanks, this looks good to me!